### ISSUE TITLE
Scale gradient of ImageNet by 1e+9

### ISSUE THREAD

**Issue by XinDongol** [jiecaoyu/XNOR-Net-PyTorch#47]
https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/26cca8cc793802f72b5ef2e5a2a823536a4abae8/ImageNet/networks/util.py#L89
`self.target_modules[index].grad.data = self.target_modules[index].grad.data.mul(1e+9)
`

I notice cifar 10 does not have this line. Is there any difference between ImageNet and Cifar. Could you please give some hints that why we want this line?

**Comment by jiecaoyu** [comment]
My assumption is that this line is used for adjusting the weight decay strength for the binary layers. For CIFAR10, we are using the number of input channels to adjust the weight decay. However, for AlexNet, we need to use a much lower weight decay and, therefore, we are using ```1e9```.

**Comment by XinDongol** [comment]
Thanks for your quick reply. 

1. I checked previous discussion on this problem:[Why the gradient in backward after the scaled sign function looks unlike the gradient in the paper?](https://github.com/jiecaoyu/XNOR-Net-PyTorch/issues/1) There are two explains.  
    
     - [used to fix the precision](https://github.com/jiecaoyu/XNOR-Net-PyTorch/issues/1#issuecomment-336533258)
     - The gradients in the layers closed to the input will be very small. When we are doing the computation of Adam, we need to calculate the average of past squared gradients. To avoid making the past squared gradients too small, we may need mul(1e+9)

2. I notice that the default weights decay for CIFAR and ImageNet are both `1e-5`. (Or maybe ImageNet is not 1e-5, I just read from your args default. If it is wrong, could you please point out the right one.) What's more, it is a little hard for me to understand the relationship between `1e+9` and weights decay. 

3. I trid to replace `adam` with `SGD`. The network will not converge at all. When I move `1e+9`, SGD can work.(But I just run a few iterations. Still not sure about final performance of SGD) So, it seems that `1e+9` is only appliable for `adam`. The source code of XNOR from author of XNOR paper: 
```Python
if opt.optimType == 'adam' then
        gradParameters:mul(1e+9);
```

**Comment by jiecaoyu** [comment]
For Adam, each gradient will be normalized. Therefore, it is okay to time the gradients with 1e9. This does not hold for SGD.

**Comment by legolas123** [comment]
I have implemented the xnornet ++ paper with resnet-18. In the quantization file that I am using, I am not doing any fancy gradient update and also multiplication by 1e9 is also not required. Basically x_out is taken to be mean_of_absolute_value(x_in)*sign(x_in). So grad_in = grad_out*clipping_grad_due_to_sign. Also xnornet++ is improvement over xnornet and my implementation reaches 60% accuracy for 1 bit activation and 1 bit width.
The link is https://github.com/legolas123/cv-tricks.com/tree/master/xnornet_plusplus
